crates/triangulation/src/intersection.rs

```diff
@@ -0,0 +1,424 @@
+use crate::point;
+use std::cmp::Ordering;
+use std::collections::{BTreeMap, HashSet};
+use std::hash::Hash;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct Event {
+    pub p: point::Point,
+    pub index: point::Index,
+    pub is_top: bool,
+}
+
+impl Event {
+    pub fn new(p: point::Point, index: point::Index, is_top: bool) -> Self {
+        Self { p, index, is_top }
+    }
+}
+
+impl PartialOrd for Event {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Event {
+    fn cmp(&self, other: &Self) -> Ordering {
+        if self.p == other.p {
+            if self.is_top == other.is_top {
+                self.index.cmp(&other.index)
+            } else {
+                // Note the reversed comparison for is_top
+                other.is_top.cmp(&self.is_top)
+            }
+        } else {
+            // Assuming Point implements PartialOrd
+            self.p.cmp(&other.p)
+        }
+    }
+}
+
+#[derive(Default, Clone)]
+struct EventData {
+    tops: Vec<usize>,
+    bottoms: Vec<usize>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct OrderedPair(point::Index, point::Index);
```

Comment:
Do the fields here need to be public for this to be useful? If so, I would also create accessors so you can't instantiate an invalid `OrderedPair` (i.e. without calling `new`. You can also encapsulate this in a module to prevent invalid construction even within this module, for example.

Reply:
Yes, I think that's correct.

The visibility is likewise confusing to me, but I believe these fields will be private outside this module. So if you have an `OrderedPair` elsewhere you would not be able to access `ordered_pair.0` or `ordered_pair.1`. I think this is what you want, as it also prevents instantiating like `OrderedPair(10, 5)` and forces construction via `::new` where the order is determined. Then you'd add accessors to be able to use the properties but maintain invariant.